# Импорт используемых библиотек



In [ ]:
import pandas as pd
import numpy as np #для матричных вычислений
import pandas as pd #для анализа и предобработки данных
from sklearn import model_selection
from sklearn import linear_model #линейные модели
from sklearn import metrics #метрики
from sklearn import preprocessing #предобработка
from sklearn.model_selection import train_test_split #разделение выборки
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.neural_network import MLPRegressor
import tensorflow as tf
from keras.models import Sequential
from keras.layers import Dense, LSTM

# Расследовательный анализ данных

## Чтение предоставленных данных

In [ ]:
train_data = pd.read_excel('data/Обучающая выборка.xlsx')

In [ ]:
test_data = pd.read_excel('data/Тестовая выборка.xlsx')

In [ ]:
validation_data = pd.read_excel('data/контрольная выборка.xlsx')

## Просмотр обучающей выборки

In [ ]:
train_data.head()

Параметр 1  \
0  Частота вращения ротора высокого давления   
1                                         N2   
2                                    об/мин.   
3                                       8944   
4                                       9043   

                                 Параметр 2  \
0  Частота вращения ротора низкого давления   
1                                        N1   
2                                   об/мин.   
3                                      3491   
4                                      3595   

                                          Параметр 3  \
0  Давление воздуха за компрессором, избыточное, ...   
1                                                 P2   
2                                            кгc/см2   
3                                               2.19   
4                                                2.3   

                         Параметр 4  \
0  Температура атмосферного воздуха   
1                              Tatm   
2                     град. цельсия   
3                               -20   
4                               -20   

                                     Параметр 5  \
0  Температура воздуха на входе в газогенератор   
1                                            T1   
2                                 град. цельсия   
3                                           -19   
4                                         -19.6   

                                Параметр 6                     Параметр 7  \
0  Частота вращения ротора силовой турбины  Давление атмосферного воздуха   
1                                       N3                             Ba   
2                                  об/мин.                      мм.рт.ст.   
3                                     1315                            746   
4                                     1374                            746   

                                          Параметр 8  \
0  Расход топливного газа, без корректировки на о...   
1                                                Qtg   
2                                            нм3/час   
3                                                836   
4                                                844   

                              Параметр 9  \
0  Крутящий момент на промежуточном валу   
1                                    Мкр   
2                                    Н*м   
3                                   2040   
4                                   2246   

                                         Параметр 10  ...  \
0  Достоверная относительная влажность атмосферно...  ...   
1                                                 RH  ...   
2                                                  %  ...   
3                                              63.37  ...   
4                                              63.62  ...   

                                        Параметр 55  \
0  Гармоника вращения рессоры привода КПА двигателя   
1                                              Fкпа   
2                                              мм/с   
3                                              0.26   
4                                              0.35   

                                         Параметр 56  \
0  Гармоника вращения центробежного суфлера двига...   
1                                                Fцс   
2                                               мм/с   
3                                               0.04   
4                                               0.02   

                                         Параметр 57  \
0  Гармоника вращ. сепаратора подшипника передней...   
1                                                Fc4   
2                                               мм/с   
3                                               0.02   
4                                               0.02   

                             Параметр 58  \
0  2-я гармоника ротора низкого давления   
1                                    2F1   
2             

In [ ]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 143167 entries, 0 to 143166
Data columns (total 64 columns):
 #   Column             Non-Null Count   Dtype 
---  ------             --------------   ----- 
 0   Параметр 1         143167 non-null  object
 1   Параметр 2         143167 non-null  object
 2   Параметр 3         143167 non-null  object
 3   Параметр 4         143167 non-null  object
 4   Параметр 5         143167 non-null  object
 5   Параметр 6         143167 non-null  object
 6   Параметр 7         143167 non-null  object
 7   Параметр 8         143167 non-null  object
 8   Параметр 9         143167 non-null  object
 9   Параметр 10        143167 non-null  object
 10  Параметр 11        143166 non-null  object
 11  Параметр 12        143166 non-null  object
 12  Параметр 13        143167 non-null  object
 13  Параметр 14        143167 non-null  object
 14  Параметр 15        143167 non-null  object
 15  Параметр 16        143167 non-null  object
 16  Параметр 17        1

In [ ]:
print(f'Размер обучающей выборки: {train_data.shape}')

Размер обучающей выборки: (143167, 64)


## Просмотр тестовой выборки






In [ ]:
test_data.head()

Параметр 1  \
0  Частота вращения ротора высокого давления   
1                                         N2   
2                                    об/мин.   
3                                      10382   
4                                      10377   

                                 Параметр 2  \
0  Частота вращения ротора низкого давления   
1                                        N1   
2                                   об/мин.   
3                                      7392   
4                                      7388   

                                          Параметр 3  \
0  Давление воздуха за компрессором, избыточное, ...   
1                                                 P2   
2                                            кгc/см2   
3                                               9.63   
4                                               9.63   

                         Параметр 4  \
0  Температура атмосферного воздуха   
1                              Tatm   
2                     град. цельсия   
3                             -19.9   
4                             -19.8   

                                     Параметр 5  \
0  Температура воздуха на входе в газогенератор   
1                                            T1   
2                                 град. цельсия   
3                                         -21.2   
4                                         -21.4   

                                Параметр 6                     Параметр 7  \
0  Частота вращения ротора силовой турбины  Давление атмосферного воздуха   
1                                       N3                             Ba   
2                                  об/мин.                      мм.рт.ст.   
3                                     3697                            746   
4                                     3707                            745   

                                          Параметр 8  \
0  Расход топливного газа, без корректировки на о...   
1                                                Qtg   
2                                            нм3/час   
3                                               2758   
4                                               2743   

                              Параметр 9  \
0  Крутящий момент на промежуточном валу   
1                                    Мкр   
2                                    Н*м   
3                                  15589   
4                                  15534   

                                         Параметр 10  ...  \
0  Достоверная относительная влажность атмосферно...  ...   
1                                                 RH  ...   
2                                                  %  ...   
3                                              65.14  ...   
4                                              65.49  ...   

                                        Параметр 55  \
0  Гармоника вращения рессоры привода КПА двигателя   
1                                              Fкпа   
2                                              мм/с   
3                                              0.69   
4                                              0.92   

                                         Параметр 56  \
0  Гармоника вращения центробежного суфлера двига...   
1                                                Fцс   
2                                               мм/с   
3                                               0.05   
4                                               0.05   

                                         Параметр 57  \
0  Гармоника вращ. сепаратора подшипника передней...   
1                                                Fc4   
2                                               мм/с   
3                                               0.04   
4                                               0.11   

                             Параметр 58  \
0  2-я гармоника ротора низкого давления   
1                                    2F1   
2             

In [ ]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 143167 entries, 0 to 143166
Data columns (total 64 columns):
 #   Column             Non-Null Count   Dtype 
---  ------             --------------   ----- 
 0   Параметр 1         143167 non-null  object
 1   Параметр 2         143167 non-null  object
 2   Параметр 3         143167 non-null  object
 3   Параметр 4         143167 non-null  object
 4   Параметр 5         143167 non-null  object
 5   Параметр 6         143167 non-null  object
 6   Параметр 7         143167 non-null  object
 7   Параметр 8         143167 non-null  object
 8   Параметр 9         143167 non-null  object
 9   Параметр 10        143167 non-null  object
 10  Параметр 11        143166 non-null  object
 11  Параметр 12        143166 non-null  object
 12  Параметр 13        143167 non-null  object
 13  Параметр 14        143167 non-null  object
 14  Параметр 15        143167 non-null  object
 15  Параметр 16        143167 non-null  object
 16  Параметр 17        1

In [ ]:
print(f'Размер тестовой выборки: {test_data.shape}')

Размер тестовой выборки: (17900, 64)


## Просмотр контрольной выборки


Только в контрольной выборке нету целевого признака

In [ ]:
validation_data.head()

Параметр 1  \
0  Частота вращения ротора высокого давления   
1                                         N2   
2                                    об/мин.   
3                                      10412   
4                                      10397   

                                 Параметр 2  \
0  Частота вращения ротора низкого давления   
1                                        N1   
2                                   об/мин.   
3                                      7441   
4                                      7421   

                                          Параметр 3  \
0  Давление воздуха за компрессором, избыточное, ...   
1                                                 P2   
2                                            кгc/см2   
3                                                9.8   
4                                               9.71   

                         Параметр 4  \
0  Температура атмосферного воздуха   
1                              Tatm   
2                     град. цельсия   
3                               -20   
4                             -19.9   

                                     Параметр 5  \
0  Температура воздуха на входе в газогенератор   
1                                            T1   
2                                 град. цельсия   
3                                         -20.9   
4                                         -21.1   

                                Параметр 6                     Параметр 7  \
0  Частота вращения ротора силовой турбины  Давление атмосферного воздуха   
1                                       N3                             Ba   
2                                  об/мин.                      мм.рт.ст.   
3                                     3706                            746   
4                                     3706                            746   

                                          Параметр 8  \
0  Расход топливного газа, без корректировки на о...   
1                                                Qtg   
2                                            нм3/час   
3                                               2823   
4                                               2797   

                              Параметр 9  \
0  Крутящий момент на промежуточном валу   
1                                    Мкр   
2                                    Н*м   
3                                  15945   
4                                  15786   

                                         Параметр 10  ...  \
0  Достоверная относительная влажность атмосферно...  ...   
1                                                 RH  ...   
2                                                  %  ...   
3                                              64.49  ...   
4                                              64.81  ...   

                              Параметр 54  \
0  2-я гармоника ротора высокого давления   
1                                     2F2   
2                                    мм/с   
3                                    0.27   
4                                    0.48   

                                        Параметр 55  \
0  Гармоника вращения рессоры привода КПА двигателя   
1                                              Fкпа   
2                                              мм/с   
3                                              0.58   
4                                              0.68   

                                         Параметр 56  \
0  Гармоника вращения центробежного суфлера двига...   
1                                                Fцс   
2                                               мм/с   
3                                               0.05   
4                                               0.06   

                                         Параметр 57  \
0  Гармоника вращ. сепаратора подшипника передней...   
1                                                Fc4   
2                                           

In [ ]:
validation_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17869 entries, 0 to 17868
Data columns (total 63 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Параметр 1   17869 non-null  float64
 1   Параметр 2   17869 non-null  float64
 2   Параметр 3   17869 non-null  float64
 3   Параметр 4   17869 non-null  float64
 4   Параметр 5   17869 non-null  float64
 5   Параметр 6   17869 non-null  float64
 6   Параметр 7   17869 non-null  float64
 7   Параметр 8   17869 non-null  float64
 8   Параметр 9   17869 non-null  float64
 9   Параметр 10  17869 non-null  float64
 10  Параметр 11  17869 non-null  float64
 11  Параметр 12  17869 non-null  float64
 12  Параметр 13  17869 non-null  float64
 13  Параметр 14  17869 non-null  float64
 14  Параметр 15  17869 non-null  float64
 15  Параметр 16  17869 non-null  float64
 16  Параметр 17  17869 non-null  float64
 17  Параметр 18  17869 non-null  float64
 18  Параметр 19  17869 non-null  float64
 19  Пара

In [ ]:
print(f'Размер контрольной выборки: {validation_data.shape}')

Размер контрольной выборки: (17869, 63)


## Генерация отчета по данным

In [ ]:
from ydata_profiling import ProfileReport
profile = ProfileReport(train_data)
profile.to_file("data ProfileReport.html")

 Вывод: после просмотра графиков распределения всех параметров аномальных значений не наблюдается, также в выборках нету пропуков. Чтобы подготовить данные для обучения нужно удалить текстовые строки (0, 1, 2) и изменить тип данных в выборках на float вместо object. Также нужно исправить название признака 'Параметр 10.1' на 'Параметр 12' в тестовой и контрольной выборке.

## Подготовка данных

In [ ]:
#Удаляем текстовые строки
train_data = train_data.drop([0, 1, 2])
test_data = test_data.drop([0, 1, 2])
validation_data = validation_data.drop([0, 1, 2])

#Обновляем индексы в датафрейме
train_data = train_data.reset_index(drop=True)
test_data = test_data.reset_index(drop=True)
validation_data = validation_data.reset_index(drop=True)

#Меняем тип данных object на float
train_data = train_data.apply(lambda x: x.astype(float) if x.dtype == 'object' else x)
test_data = test_data.apply(lambda x: x.astype(float) if x.dtype == 'object' else x)
validation_data = validation_data.apply(lambda x: x.astype(float) if x.dtype == 'object' else x)

#Меняем название столбца 'Параметр 10.1'в test_data и validation_data на 'Параметр 12'
test_data.rename(columns={'Параметр 10.1': 'Параметр 12'}, inplace=True)
validation_data.rename(columns={'Параметр 10.1': 'Параметр 12'}, inplace=True)

Тестовые данные включают в себя 143167 строки и 64 признака, включая "целевой".

#Построение моделей

Подготовим обучающие и тестовые выборки для обучения

In [ ]:
features = train_data.drop('Выходной параметр', axis=1).columns
X_train = train_data[features]
y_train = train_data['Выходной параметр']
X_test = test_data[features]
y_test = test_data['Выходной параметр']

## MLP (Многослойный персептрон)

In [ ]:
mlp = MLPRegressor()
mlp.fit(X_train, y_train)
#прогнозы для новых данных
y_train_pred_mlp = mlp.predict(X_train)
y_test_pred_mlp = mlp.predict(X_test)


### Предсказанные значения на валидационной выборке для RNN



In [ ]:
y_val_pred_mlp = mlp.predict(validation_data)

df = pd.DataFrame(y_val_pred_mlp)
df.to_excel('y_val_pred_mlp.xlsx', index=False)



### Вычисление метрик MAE и MAPE

In [ ]:
print(f'Значение метрики MAE на обучающих данных:{round(metrics.mean_absolute_error(y_train_pred_mlp, y_train), 3)}')
print(f'Значение метрики MAPE на обучающих данных:{round(metrics.mean_absolute_percentage_error(y_train_pred_mlp, y_train), 3)}')
print('')
print(f'Значение метрики MAE на тестовых данных:{round(metrics.mean_absolute_error(y_test_pred_mlp, y_test), 3)}')
print(f'Значение метрики MAPE на тестовых данных:{round(metrics.mean_absolute_percentage_error(y_test_pred_mlp, y_test), 3)}')


Значение метрики MAE на обучающих данных:5.333
Значение метрики MAPE на обучающих данных:0.008

Значение метрики MAE на тестовых данных:5.347
Значение метрики MAPE на тестовых данных:0.008


Модель линейной регрессии очень точно предсказывает значение целевого признака. Переобучения модели не наблюдается.

### Проведем нормализацию данных, чтобы попробовать улучшить показатели метрики

In [ ]:
scaler = preprocessing.MinMaxScaler()
#Вычисляем параметры для нормализации - min и max для каждого столбца
scaler.fit(X_train)
#Производим преобразование для каждой из выборок
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [ ]:
mlp_scaler = MLPRegressor()
mlp_scaler.fit(X_train_scaled, y_train)
#Делаем предсказание для тренировочной выборки
y_train_scaler_pred = mlp_scaler.predict(X_train_scaled)
#Делаем предсказание на тестовой выборке
y_test_scaler_pred = mlp_scaler.predict(X_test_scaled)

/usr/local/lib/python3.10/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


### Предсказанные значения на валидационной выборке для модели линейной после проведения нормализации


In [ ]:
X_val_scaled = scaler.transform(validation_data)
y_val_pred_mlp_sc = mlp_scaler.predict(X_val_scaled)
df = pd.DataFrame(y_val_pred_mlp_sc)
df.to_excel('y_val_pred_mlp_sc.xlsx', index=False)

### Вычисление метрик

In [ ]:
print(f'Значение метрики MAE на обучающих данных:{round(metrics.mean_absolute_error(y_train_scaler_pred, y_train), 3)}')
print(f'Значение метрики MAPE на обучающих данных:{round(metrics.mean_absolute_percentage_error(y_train_scaler_pred, y_train), 3)}')
print('')
print(f'Значение метрики MAE на тестовых данных:{round(metrics.mean_absolute_error(y_test_scaler_pred, y_test), 3)}')
print(f'Значение метрики MAPE на тестовых данных:{round(metrics.mean_absolute_percentage_error(y_test_scaler_pred, y_test), 3)}')

Значение метрики MAE на обучающих данных:1.101
Значение метрики MAPE на обучающих данных:0.002

Значение метрики MAE на тестовых данных:1.519
Значение метрики MAPE на тестовых данных:0.002


### Нормализация данных улучшила метрики моделей

## RNN (Рекуррентные нейронные сети)

In [ ]:
# Создание модели RNN
rnn = Sequential()
rnn.add(LSTM(100, return_sequences=True, input_shape=(None, 1)))
rnn.add(LSTM(100))
rnn.add(Dense(1))
# Компиляция
rnn.compile(loss='mean_squared_error', optimizer='adam')
# Обучение модели
rnn.fit(X_train, y_train)
# Предсказания на данных
y_train_pred_rnn = rnn.predict(X_train)
y_test_pred_rnn = rnn.predict(X_test)

/usr/local/lib/python3.10/dist-packages/keras/src/layers/rnn/rnn.py:204: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


4474/4474 ━━━━━━━━━━━━━━━━━━━━ 475s 105ms/step - loss: 341466.8750
4474/4474 ━━━━━━━━━━━━━━━━━━━━ 164s 37ms/step
560/560 ━━━━━━━━━━━━━━━━━━━━ 21s 38ms/step


### Предсказанные значения на валидационной выборке для RNN

In [ ]:
y_val_pred_rnn = rnn.predict(validation_data)
df = pd.DataFrame(y_val_pred_rnn)
df.to_excel('y_val_pred_rnn.xlsx', index=False)

559/559 ━━━━━━━━━━━━━━━━━━━━ 20s 36ms/step


### Вычисление метрик

In [ ]:
print(f'Значение метрики MAE на обучающих данных:{round(metrics.mean_absolute_error(y_train_pred_rnn, y_train), 3)}')
print(f'Значение метрики MAPE на обучающих данных:{round(metrics.mean_absolute_percentage_error(y_train_pred_rnn, y_train), 3)}')
print('')
print(f'Значение метрики MAE на тестовых данных:{round(metrics.mean_absolute_error(y_test_pred_rnn, y_test), 3)}')
print(f'Значение метрики MAPE на тестовых данных:{round(metrics.mean_absolute_percentage_error(y_test_pred_rnn, y_test), 3)}')

Значение метрики MAE на обучающих данных:290.175
Значение метрики MAPE на обучающих данных:0.716

Значение метрики MAE на тестовых данных:296.119
Значение метрики MAPE на тестовых данных:0.731


### Обучение и измерения метрик на нормализованных данных

In [ ]:
rnn_scaler = Sequential()
rnn_scaler.add(LSTM(100, return_sequences=True, input_shape=(None, 1)))
rnn_scaler.add(LSTM(100))
rnn_scaler.add(Dense(1))
rnn_scaler.compile(loss='mean_squared_error', optimizer='adam')
rnn_scaler.fit(X_train_scaled, y_train)
#Делаем предсказание для тренировочной выборки
y_train_scaler_pred = rnn_scaler.predict(X_train_scaled)
#Делаем предсказание на тестовой выборке
y_test_scaler_pred = rnn_scaler.predict(X_test_scaled)

4474/4474 ━━━━━━━━━━━━━━━━━━━━ 439s 97ms/step - loss: 340565.8125
4474/4474 ━━━━━━━━━━━━━━━━━━━━ 153s 34ms/step
560/560 ━━━━━━━━━━━━━━━━━━━━ 20s 35ms/step


In [ ]:
y_val_pred_rnn_scaler = rnn_scaler.predict(validation_data)
df = pd.DataFrame(y_val_pred_rnn_scaler)
df.to_excel('y_val_pred_rnn_scaler.xlsx', index=False)

559/559 ━━━━━━━━━━━━━━━━━━━━ 19s 34ms/step


### Вычисление метрик

In [ ]:
print(f'Значение метрики MAE на обучающих данных:{round(metrics.mean_absolute_error(y_train_scaler_pred, y_train), 3)}')
print(f'Значение метрики MAPE на обучающих данных:{round(metrics.mean_absolute_percentage_error(y_train_scaler_pred, y_train), 3)}')
print('')
print(f'Значение метрики MAE на тестовых данных:{round(metrics.mean_absolute_error(y_test_scaler_pred, y_test), 3)}')
print(f'Значение метрики MAPE на тестовых данных:{round(metrics.mean_absolute_percentage_error(y_test_scaler_pred, y_test), 3)}')

Значение метрики MAE на обучающих данных:286.76
Значение метрики MAPE на обучающих данных:0.702

Значение метрики MAE на тестовых данных:292.705
Значение метрики MAPE на тестовых данных:0.717


### Наблюдения:
1) Лучшие показатели по метрикам получились и модели MLP с нормализованными данными

2) Для модели RNN получись метрики намного хуже, чем для MLP

3) Нормализация данных улучшает качество модели

4) Нейросеть дольше обучается на нормализованных данных, чем на обычных